rpc/codec/server.go

```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// For requests with a RequestHeader, we compare the header.User against
+// the client certificate Subject.CommonName.
+// For other requests, we either allow all (test requests), or
+// require the "node user".
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	var err error
+	var requestedUser string
+
+	switch request := request.(type) {
+	case cockroach_proto.Request:
```

Comment:
Rather than use `cockroach_proto.Request` here, I think you could define your own more restrictive interface:

```
type userRequest interface {
  GetUser() string
}
```

Then you could fix the `{Echo,Arith}Request` messages by adding a user field. Ditto for `{Ping,Gossip,RaftMessage}Request`.